### PR TITLE
Add DotNet style guide basic layout

### PR DESCRIPTION
**Pull request by matiasbeckerle**
## Background

We were discussing a lot in several pull requests across different projects. This PR aims to set a base for putting the conclusions about style stuff. Please feel free to add, modify, delete, suggest, comment, whatever.

## Changes

- Added a new folder and readme file for `DotNet`.
- Added a basic layout for specifying a simple rule (use of var).
- Added reference in the root readme.

## Pending

- All the rules. LOL

## Notes

- As you may see, this is only a layout proposal so feel free to comment about it. Looked for some other references but I couldn't find any nice markdown based on C# (even in _awesome_ lists) so I started a new one. Maybe we got the chance of having lot of forks, who knows!
- Once I got the approval for this one, rules will come in (slowly).
- I will add a couple of reviewers but feel free to add anyone!

### CODE REVIEW

File: DotNet/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+HttpClient httpClient = new HttpClient();
+
+// Good
+var httpClient = new HttpClient();
+
+// Allowed for using an interface instead of a class
+IUserService userService = new UserService();
+
+// Allowed because the variable is poorly named from legacy code
+Graph g = new Graph();
```
Comment (andresmoschini) [inline DotNet/README.md]: I think that this scenario could be removed, it is normal to break the rules in exceptional cases like legacy code.

Comment (andresmoschini) [inline DotNet/README.md]: BTW, I would like to see an example of this Eric Lippert's sentence : 

> Use explicit types if doing so is necessary for the code to be correctly understood and maintained. 

Personally, I prefer to _always_ use `var`.

Comment (martinkeimel) [inline DotNet/README.md]: +1. Each team should decide if legacy code should be adjusted (gradually or all at once) or not depending on each project's context

**Comment by andresmoschini** [comment]
Nice initiative @matiasbeckerle, thanks!

I want to mention that VS2017 supports some _editorconfig_ extensions that could be useful to document and encourage the usage of some of the guidelines. See more details in:
* [.NET Code Style Settings For EditorConfig](https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-code-style-settings-reference)
* [Roslyn .editorconfig](https://github.com/dotnet/roslyn/blob/master/.editorconfig)
* [Roslyn .editorconfig Pull Request](https://github.com/dotnet/roslyn/pull/15029)

**Comment by mvazquez-ms** [comment]
Nice! We really need this.

The only comment I have so far is I think we should name it **C#** style-guide. DotNet is too broad, and if we put anything DotNet-related here, it will become huge. (e.g. On 3.0, we had tons of rules for XAML code, that's DotNet)

File: DotNet/README.md (hunk 1 above)
Comment (matiasbeckerle) [inline DotNet/README.md]: Good point, will remove.

Comment (matiasbeckerle) [inline DotNet/README.md]: 😄 Agree. Maybe if you try a comment on the blogpost, who knows... Eric seems to leave really cool responses as @mvazquez-ms points every day.

**Comment by matiasbeckerle** [comment]
Agree with you @mvazquez-ms. Actually I forgot to put that question on the PR description, as usually forgot something. I will rename this one. Thanks.

**Comment by matiasbeckerle** [comment]
Thank you @andresmoschini for the `.editorconfig` thing. I also believe we will have a resharper file too (as I told you today).

**Comment by matiasbeckerle** [comment]
Thanks @AlphaGit, appreciated. Agree on both things:

- I will move those below.
- I will rename the title, it only was a placeholder. BTW I'm the worst grouping stuff so I will need naming ideas in the near future.

**Comment by matiasbeckerle** [comment]
Thank you all for participating, commenting, smiling or thumbing up! Appreciated. Feel free to join in upcoming discussions!

File: DotNet/README.md (hunk 1 above)
Comment (mvazquez-ms) [inline DotNet/README.md]: > BTW, I would like to see an example of this Eric Lippert's sentence:

I have one example in mind where explicit type is _slightly_ more readable than var. (For the record, I prefer `var` 99.99% of the time)

A method receiving a `Func<T1,T2,TResult>`

```
public void SomeMethod(Func<int, int, bool> intComparer)
{
    // do something with intComparer
}
```

```
// Explicit Type
Func<int, int, bool> greaterThanComparer = (x, y) => x > y;
SomeMethod(greaterThanComparer);

// Implicit Type 1 (this won't compile as type cannot be inferred)
var lessThanComparer = (x, y) => x < y;
SomeMethod(lessThanComparer);

// Implicit Type 2 (I find this harder to read than the explicit type version)
var lessThanComparer = new Func<int, int, bool>((x, y) => x < y);
SomeMethod(lessThanComparer);

// Local Function (unusual syntax)
bool equalityComparer(int x, int y) => x == y;
SomeMethod(equalityComparer);
```

Of course this is a rare scenario, and it's OK if someone still prefers the implicit type version..

Comment (andresmoschini) [inline DotNet/README.md]: Thanks @mvazquez-ms!

But, hehe, I still prefer the implicit version:

```csharp
void SomeMethod(Func<int, int, bool> intComparer)
{
    // do something with intComparer
}

void Main()
{
    SomeMethod((x, y) => x > y);
}
```

**Pull request merged**